lib/src/command_runner.dart

```diff
@@ -180,8 +180,11 @@ class PubCommandRunner extends CommandRunner {
       if (topLevelResults['trace']) {
         log.dumpTranscript();
       } else if (!isUserFacingException(error)) {
-        // TODO(23505): Implement proper shell escaping, not a partial hack.
-        protectArgument(String x) => x.contains(' ') ? '"$x"' : x;
+        // Escape the argument for users to copy-paste in bash.
+        protectArgument(String x) =>
+            RegExp(r'^[a-zA-Z0-9-_]+$').stringMatch(x) == null
+                ? "'${x.replaceAll("'", "'\\''")}'"
```

Comment:
You cannot escape a single quote inside a single quote:
https://www.gnu.org/software/bash/manual/html_node/Single-Quotes.html#Single-Quotes
You probably need to use double quotes and escape all of ‘$’, ‘`’, ‘\’

Reply:
Not really... Read carefully I escape single quote, by ending single quotation then having an escaped single quote and starting single quotation again :)

If there is no space between end and start of single quotation it doesn't become a new argument...

Reply:
Explain this in a comment ...